juju/juju
# Relocate Leadership Pinning for Upgrade-Series

## Description of change

This patch does the following: 
- Changes the new leadership API for pinning manipulations to use the authenticated caller as the entity for pin/unpin operations.
- Relocates the API and API server used for this to the common packages.
- Adds the leadership API as a constituent of the upgrade-series API.
- Adds pin/unpin logic to the upgrade-series worker.
- Removes pin/unpin logic from the upgrade-series client command.

Note that the pin/unpin feedback no longer appears in the upgrade-series command stdout. This is now handled by the worker and thus appears in the machine logs.

## QA steps

- Bootstrap to LXD.
- `juju deploy redis`.
- `juju add-unit redis`.
- `juju upgrade-series prepare 0` and confirm.
- `juju upgrade-series prepare 1` and confirm.
- The leader should be redis/0. Run `upgrade-series complete 0`
- Stop the container for machine 0 with `lxc stop {container ID}`.
- Watch `juju status` for a while to ensure that leadership does not change from redis/0, even though the machine is down.
- `juju upgrade-series complete 1`
- After a moment, leadership should change to redis/1.

## Documentation changes

None.

## Bug reference

N/A
